src/NuGetPackageVerifier/Rules/PackageTypesRule.cs

```diff
@@ -0,0 +1,37 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+
+namespace NuGetPackageVerifier.Rules
+{
+    public class PackageTypesRule : IPackageVerifierRule
+    {
+        public IEnumerable<PackageVerifierIssue> Validate(PackageAnalysisContext context)
+        {
+            var discoveredTypes = new HashSet<string>(context.Metadata.PackageTypes.Select(t=>t.Name));
+            var expectedTypes = new HashSet<string>(context.Options.PackageTypes ?? Enumerable.Empty<string>());
+
+            var intersection = new HashSet<string>(expectedTypes.Intersect(discoveredTypes));
```

Comment:
You don't have to do it but I think this code can be simplified by doing something like,
```c#
var missingTypes = expectedTypes.Except(discoveredTypes);
var unexpectedTypes = discoveredTypes.Except(expectedTypes);
```